@dotnet/dnceng Helix error:
```

.packages/microsoft.dotnet.helix.sdk/7.0.0-beta.22179.1/tools/xharness-runner/XHarnessRunner.targets(122,5): error MSB4018: The "CreateXHarnessAppleWorkItems" task failed unexpectedly.
System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
 ---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid because of errors in the certificate chain: RevocationStatusUnknown
```

--- a/src/libraries/Common/src/Interop/Windows/Advapi32/Interop.ServiceProcessOptions.cs
+++ b/src/libraries/Common/src/Interop/Windows/Advapi32/Interop.ServiceProcessOptions.cs
@@ -95,14 +95,14 @@ internal static partial class Interop
             internal const int SERVICE_INTERACTIVE_PROCESS = 0x00000100;
             //internal const int SERVICE_PKG_PROCESS = 0x00000200;
 
-            internal const int SERVICE_TYPE_ALL =
-                SERVICE_WIN32 |
-                SERVICE_ADAPTER |
-                SERVICE_DRIVER |
-                SERVICE_INTERACTIVE_PROCESS;
-              //SERVICE_USER_SERVICE |
-              //SERVICE_USERSERVICE_INSTANCE |
-              //SERVICE_PKG_SERVICE;
+            // internal const int SERVICE_TYPE_ALL =
+            //    SERVICE_WIN32 |
+            //    SERVICE_ADAPTER |
+            //    SERVICE_DRIVER |
+            //    SERVICE_INTERACTIVE_PROCESS;
+            //    SERVICE_USER_SERVICE |
+            //    SERVICE_USERSERVICE_INSTANCE |
+            //    SERVICE_PKG_SERVICE;
         }
 
         internal static partial class ServiceAccessOptions

--- a/src/libraries/Common/src/Interop/Windows/Advapi32/Interop.ServiceProcessOptions.cs
+++ b/src/libraries/Common/src/Interop/Windows/Advapi32/Interop.ServiceProcessOptions.cs
@@ -82,27 +82,7 @@ internal static partial class Interop
                 SERVICE_WIN32_OWN_PROCESS |
                 SERVICE_WIN32_SHARE_PROCESS;
 
-            //internal const int SERVICE_USER_SERVICE = 0x00000040;
-            //internal const int SERVICE_USERSERVICE_INSTANCE = 0x00000080;
-
-            //internal const int SERVICE_USER_SHARE_PROCESS =
-            //    SERVICE_USER_SERVICE |
-            //    SERVICE_WIN32_SHARE_PROCESS;
-            //internal const int SERVICE_USER_OWN_PROCESS =
-            //    SERVICE_USER_SERVICE |
-            //    SERVICE_WIN32_OWN_PROCESS;
-
             internal const int SERVICE_INTERACTIVE_PROCESS = 0x00000100;
-            //internal const int SERVICE_PKG_PROCESS = 0x00000200;
-
-            // internal const int SERVICE_TYPE_ALL =
-            //    SERVICE_WIN32 |
-            //    SERVICE_ADAPTER |
-            //    SERVICE_DRIVER |
-            //    SERVICE_INTERACTIVE_PROCESS;
-            //    SERVICE_USER_SERVICE |
-            //    SERVICE_USERSERVICE_INSTANCE |
-            //    SERVICE_PKG_SERVICE;
         }
 
         internal static partial class ServiceAccessOptions

--- a/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs
+++ b/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs
@@ -871,7 +871,7 @@ namespace System.ServiceProcess
 
                 if (argCount > 0)
                 {
-                    char** argsAsPtr = (char**)argPointer.ToPointer();
+                    char** argsAsPtr = (char**)argPointer;
 
                     // The first arg is always the service name. We don't want to pass that in,
                     // but we can use it to set the service name on ourselves if we don't already know it.

--- a/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs
+++ b/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs
@@ -945,7 +945,8 @@ namespace System.ServiceProcess
                 statusOK = SetServiceStatus(_statusHandle, pStatus);
                 if (!statusOK)
                 {
-                    WriteLogEntry(SR.Format(SR.StartFailed, new Win32Exception().Message), EventLogEntryType.Error);
+                    string errorMessage = new Win32Exception().Message;
+                    WriteLogEntry(SR.Format(SR.StartFailed, errorMessage), EventLogEntryType.Error);
                     _status.currentState = ServiceControlStatus.STATE_STOPPED;
                     SetServiceStatus(_statusHandle, pStatus);
                 }

--- a/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs
+++ b/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs
@@ -19,6 +19,11 @@ namespace System.ServiceProcess
         private string _machineName; // Never null
         private const string DefaultMachineName = ".";
 
+        // Note that ServiceType currently does not include all of SERVICE_TYPE_ALL; see see Interop.Advapi32.ServiceTypeOptions
+        private const int AllServiceTypes = (int)(ServiceType.Adapter | ServiceType.FileSystemDriver | ServiceType.InteractiveProcess |
+                                                  ServiceType.KernelDriver | ServiceType.RecognizerDriver | ServiceType.Win32OwnProcess |
+                                                  ServiceType.Win32ShareProcess);
+
         private string? _name;
         private string? _eitherName;
         private string? _displayName;
@@ -37,7 +42,7 @@ namespace System.ServiceProcess
         public ServiceController()
         {
             _machineName = DefaultMachineName;
-            _type = Interop.Advapi32.ServiceTypeOptions.SERVICE_TYPE_ALL;
+            _type = AllServiceTypes;
         }
 
         /// <summary>
@@ -65,7 +70,7 @@ namespace System.ServiceProcess
 
             _machineName = machineName;
             _eitherName = name;
-            _type = Interop.Advapi32.ServiceTypeOptions.SERVICE_TYPE_ALL;
+            _type = AllServiceTypes;
         }
 
         private ServiceController(string machineName, Interop.Advapi32.ENUM_SERVICE_STATUS status)
@@ -310,7 +315,7 @@ namespace System.ServiceProcess
                 {
                     success = Interop.Advapi32.QueryServiceConfig(serviceHandle, bufPtr, bytesNeeded, out bytesNeeded);
                     if (!success)
-                        throw new Win32Exception(Marshal.GetLastWin32Error());
+                        throw new Win32Exception();
 
                     Interop.Advapi32.QUERY_SERVICE_CONFIG config = new Interop.Advapi32.QUERY_SERVICE_CONFIG();
                     Marshal.PtrToStructure(bufPtr, config);
@@ -389,7 +394,7 @@ namespace System.ServiceProcess
                 {
                     success = Interop.Advapi32.QueryServiceConfig(serviceHandle, bufPtr, bytesNeeded, out bytesNeeded);
                     if (!success)
-                        throw new Win32Exception(Marshal.GetLastWin32Error());
+                        throw new Win32Exception();
 
                     Interop.Advapi32.QUERY_SERVICE_CONFIG config = new Interop.Advapi32.QUERY_SERVICE_CONFIG();
                     Marshal.PtrToStructure(bufPtr, config);
@@ -467,7 +472,7 @@ namespace System.ServiceProcess
 
             _statusGenerated = false;
             _startTypeInitialized = false;
-            _type = Interop.Advapi32.ServiceTypeOptions.SERVICE_TYPE_ALL;
+            _type = AllServiceTypes;
         }
 
         /// <summary>
@@ -491,7 +496,7 @@ namespace System.ServiceProcess
             Interop.Advapi32.SERVICE_STATUS svcStatus = default;
             bool success = Interop.Advapi32.QueryServiceStatus(serviceHandle, &svcStatus);
             if (!success)
-                throw new Win32Exception(Marshal.GetLastWin32Error());
+                throw new Win32Exception();
 
             _commandsAccepted = svcStatus.controlsAccepted;
             _status = (ServiceControllerStatus)svcStatus.currentState;
@@ -632,7 +637,7 @@ namespace System.ServiceProcess
 
             if (databaseHandle.IsInvalid)
             {
-                Exception inner = new Win32Exception(Marshal.GetLastWin32Error());
+                Exception inner = new Win32Exception();
                 throw new InvalidOperationException(SR.Format(SR.OpenSC, machineName), inner);
             }
 
@@ -684,7 +689,7 @@ namespace System.ServiceProcess
             var serviceHandle = new SafeServiceHandle(Interop.Advapi32.OpenService(_serviceManagerHandle, ServiceName, desiredAccess));
             if (serviceHandle.IsInvalid)
             {
-                Exception inner = new Win32Exception(Marshal.GetLastWin32Error());
+                Exception inner = new Win32Exception();
                 throw new InvalidOperationException(SR.Format(SR.OpenService, ServiceName, _machineName), inner);
             }
 
@@ -804,7 +809,7 @@ namespace System.ServiceProcess
 
             if (!result)
             {
-                Exception inner = new Win32Exception(Marshal.GetLastWin32Error());
+                Exception inner = new Win32Exception();
                 throw new InvalidOperationException(SR.Format(SR.PauseService, ServiceName, _machineName), inner);
             }
         }
@@ -819,7 +824,7 @@ namespace System.ServiceProcess
             bool result = Interop.Advapi32.ControlService(serviceHandle, Interop.Advapi32.ControlOptions.CONTROL_CONTINUE, &status);
             if (!result)
             {
-                Exception inner = new Win32Exception(Marshal.GetLastWin32Error());
+                Exception inner = new Win32Exception();
                 throw new InvalidOperationException(SR.Format(SR.ResumeService, ServiceName, _machineName), inner);
             }
         }
@@ -835,7 +840,7 @@ namespace System.ServiceProcess
             bool result = Interop.Advapi32.ControlService(serviceHandle, command, &status);
             if (!result)
             {
-                Exception inner = new Win32Exception(Marshal.GetLastWin32Error());
+                Exception inner = new Win32Exception();
                 throw new InvalidOperationException(SR.Format(SR.ControlService, ServiceName, MachineName), inner);
             }
         }
@@ -891,7 +896,7 @@ namespace System.ServiceProcess
                 bool result = Interop.Advapi32.StartService(serviceHandle, args.Length, argPtrsHandle.AddrOfPinnedObject());
                 if (!result)
                 {
-                    Exception inner = new Win32Exception(Marshal.GetLastWin32Error());
+                    Exception inner = new Win32Exception();
                     throw new InvalidOperationException(SR.Format(SR.CannotStart, ServiceName, _machineName), inner);
                 }
             }
@@ -952,7 +957,7 @@ namespace System.ServiceProcess
             bool result = Interop.Advapi32.ControlService(serviceHandle, Interop.Advapi32.ControlOptions.CONTROL_STOP, &status);
             if (!result)
             {
-                Exception inner = new Win32Exception(Marshal.GetLastWin32Error());
+                Exception inner = new Win32Exception();
                 throw new InvalidOperationException(SR.Format(SR.StopService, ServiceName, _machineName), inner);
             }
         }

--- a/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceType.cs
+++ b/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceType.cs
@@ -6,6 +6,7 @@ namespace System.ServiceProcess
     [Flags]
     public enum ServiceType
     {
+        // Does not currently represent all of the Win32 values; see Interop.Advapi32.ServiceTypeOptions
         Adapter = Interop.Advapi32.ServiceTypeOptions.SERVICE_ADAPTER,
         FileSystemDriver = Interop.Advapi32.ServiceTypeOptions.SERVICE_FILE_SYSTEM_DRIVER,
         InteractiveProcess = Interop.Advapi32.ServiceTypeOptions.SERVICE_INTERACTIVE_PROCESS,

--- a/src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs
+++ b/src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs
@@ -86,7 +86,10 @@ namespace System.ServiceProcess.Tests
                     ServiceCommandLine, null, IntPtr.Zero, servicesDependedOn, username, password)))
                 {
                     if (serviceHandle.IsInvalid)
-                        throw new Win32Exception($"Cannot create service '{ServiceName}' with display name '{DisplayName}'. {new Win32Exception().Message}");
+                    {
+                        string errorMessage = new Win32Exception().Message;
+                        throw new Win32Exception($"Cannot create service '{ServiceName}' with display name '{DisplayName}'. {errorMessage}");
+                    }
 
                     // A local variable in an unsafe method is already fixed -- so we don't need a "fixed { }" blocks to protect
                     // across the p/invoke calls below.
@@ -98,7 +101,10 @@ namespace System.ServiceProcess.Tests
                         bool success = Interop.Advapi32.ChangeServiceConfig2(serviceHandle, Interop.Advapi32.ServiceConfigOptions.SERVICE_CONFIG_DESCRIPTION, ref serviceDesc);
                         Marshal.FreeHGlobal(serviceDesc.description);
                         if (!success)
-                            throw new Win32Exception($"Cannot set description on '{ServiceName}' with display name '{DisplayName}'. {new Win32Exception().Message}");
+                        {
+                            string errorMessage = new Win32Exception().Message;
+                            throw new Win32Exception($"Cannot set description on '{ServiceName}' with display name '{DisplayName}'. {errorMessage}");
+                        }
                     }
 
                     // Start the service after creating it
@@ -174,17 +180,24 @@ namespace System.ServiceProcess.Tests
             using (var serviceManagerHandle = new SafeServiceHandle(Interop.Advapi32.OpenSCManager(null, null, Interop.Advapi32.ServiceControllerOptions.SC_MANAGER_ALL)))
             {
                 if (serviceManagerHandle.IsInvalid)
-                    throw new Win32Exception($"Could not open SCM. {new Win32Exception().Message}");
+                {
+                    string errorMessage = new Win32Exception().Message;
+                    throw new Win32Exception($"Could not open SCM. {errorMessage}");
+                }
 
                 using (var serviceHandle = new SafeServiceHandle(Interop.Advapi32.OpenService(serviceManagerHandle, ServiceName, Interop.Advapi32.ServiceOptions.STANDARD_RIGHTS_DELETE)))
                 {
                     if (serviceHandle.IsInvalid)
-                        throw new Win32Exception($"Could not find service '{ServiceName}'. {new Win32Exception().Message}");
+                    {
+                        string errorMessage = new Win32Exception().Message;
+                        throw new Win32Exception($"Could not find service '{ServiceName}'. {errorMessage}");
+                    }
 
                     TestService.DebugTrace("TestServiceInstaller: instructing ServiceController to Delete service " + ServiceName);
                     if (!Interop.Advapi32.DeleteService(serviceHandle))
                     {
-                        throw new Win32Exception($"Could not delete service '{ServiceName}'. {new Win32Exception().Message}");
+                        string errorMessage = new Win32Exception().Message;
+                        throw new Win32Exception($"Could not delete service '{ServiceName}'. {errorMessage}");
                     }
                 }
             }

--- a/src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs
+++ b/src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs
@@ -87,8 +87,9 @@ namespace System.ServiceProcess.Tests
                 {
                     if (serviceHandle.IsInvalid)
                     {
-                        string errorMessage = new Win32Exception().Message;
-                        throw new Win32Exception($"Cannot create service '{ServiceName}' with display name '{DisplayName}'. {errorMessage}");
+                        int errorCode = Marshal.GetLastWin32Error();
+                        string errorMessage = new Win32Exception(errorCode).Message;
+                        throw new Win32Exception(errorCode, $"Cannot create service '{ServiceName}' with display name '{DisplayName}'. {errorMessage}");
                     }
 
                     // A local variable in an unsafe method is already fixed -- so we don't need a "fixed { }" blocks to protect
@@ -102,8 +103,9 @@ namespace System.ServiceProcess.Tests
                         Marshal.FreeHGlobal(serviceDesc.description);
                         if (!success)
                         {
-                            string errorMessage = new Win32Exception().Message;
-                            throw new Win32Exception($"Cannot set description on '{ServiceName}' with display name '{DisplayName}'. {errorMessage}");
+                            int errorCode = Marshal.GetLastWin32Error();
+                            string errorMessage = new Win32Exception(errorCode).Message;
+                            throw new Win32Exception(errorCode, $"Cannot set description on '{ServiceName}' with display name '{DisplayName}'. {errorMessage}");
                         }
                     }
 
@@ -181,23 +183,26 @@ namespace System.ServiceProcess.Tests
             {
                 if (serviceManagerHandle.IsInvalid)
                 {
-                    string errorMessage = new Win32Exception().Message;
-                    throw new Win32Exception($"Could not open SCM. {errorMessage}");
+                    int errorCode = Marshal.GetLastWin32Error();
+                    string errorMessage = new Win32Exception(errorCode).Message;
+                    throw new Win32Exception(errorCode, $"Could not open SCM. {errorMessage}");
                 }
 
                 using (var serviceHandle = new SafeServiceHandle(Interop.Advapi32.OpenService(serviceManagerHandle, ServiceName, Interop.Advapi32.ServiceOptions.STANDARD_RIGHTS_DELETE)))
                 {
                     if (serviceHandle.IsInvalid)
                     {
-                        string errorMessage = new Win32Exception().Message;
-                        throw new Win32Exception($"Could not find service '{ServiceName}'. {errorMessage}");
+                        int errorCode = Marshal.GetLastWin32Error();
+                        string errorMessage = new Win32Exception(errorCode).Message;
+                        throw new Win32Exception(errorCode, $"Could not find service '{ServiceName}'. {errorMessage}");
                     }
 
                     TestService.DebugTrace("TestServiceInstaller: instructing ServiceController to Delete service " + ServiceName);
                     if (!Interop.Advapi32.DeleteService(serviceHandle))
                     {
-                        string errorMessage = new Win32Exception().Message;
-                        throw new Win32Exception($"Could not delete service '{ServiceName}'. {errorMessage}");
+                        int errorCode = Marshal.GetLastWin32Error();
+                        string errorMessage = new Win32Exception(errorCode).Message;
+                        throw new Win32Exception(errorCode, $"Could not delete service '{ServiceName}'. {errorMessage}");
                     }
                 }
             }

